File: src/Users/Infrastructure/UserRepository.php
<?php

declare(strict_types=1);

namespace App\Users\Infrastructure;

use App\Categories\Domain\CustomException;
use App\Shared\Domain\EmailAddress;
use App\Shared\Domain\InvalidEmailException;
use App\Shared\Domain\UuidValueObject;
use App\Users\Domain\User;
use App\Users\Domain\UserRepositoryInterface;
use Doctrine\ORM\EntityManagerInterface;

class UserRepository implements UserRepositoryInterface
{

    public function __construct(private EntityManagerInterface $entityManager)
    {
    }

    /**
     * @throws InvalidEmailException
     * @throws CustomException
     */
    public function findByEmail(EmailAddress $email): ?User
    {
        $doctrineUser = $this->entityManager->getRepository(DoctrineUser::class)->findOneBy(['email' => $email->value]);
        if (!$doctrineUser) {
            return null;
        }

        return User::fromPrimitives(
            $doctrineUser->id,
            $doctrineUser->email,
            $doctrineUser->password,
        );
    }

    /**
     * @throws InvalidEmailException
     * @throws CustomException
     */
    public function findById(UuidValueObject $id): ?User
    {
        $repository = $this->entityManager->getRepository(DoctrineUser::class);
        $doctrineUser = $repository->find($id->value);
        if (!$doctrineUser) {
            return null;
        }

        return User::fromPrimitives(
            $doctrineUser->id,
            $doctrineUser->email,
            $doctrineUser->password,
        );
    }

    /**
     * @throws CustomException
     * @throws InvalidEmailException
     */
    public function findAll(): array
    {
        $repository = $this->entityManager->getRepository(DoctrineUser::class);
        $doctrineUsers = $repository->findAll();

        $users = [];
        foreach ($doctrineUsers as $doctrineUser) {
            $users[] = User::fromPrimitives(
                $doctrineUser->id,
                $doctrineUser->email,
                $doctrineUser->password,
            );
        }

        return $users;
    }

    public function save(User $user): void
    {
        $doctrineUser = $this->entityManager->find(
            DoctrineUser::class,
            $user->getId()->value
        ) ?? new DoctrineUser();

        $doctrineUser->id = $user->getId()->value;
        $doctrineUser->email = $user->getEmail()->value;
        $doctrineUser->password = $user->getPassword()->value;

        $this->entityManager->persist($doctrineUser);
        $this->entityManager->flush();
    }
}

File: src/main/kotlin/com/zaleort/second_brain/Users/Infrastructure/Repository/UserRepository.kt
package com.zaleort.second_brain.Users.Infrastructure.Repository

import com.zaleort.second_brain.Users.Domain.User.User
import com.zaleort.second_brain.Users.Domain.User.UserEmail
import com.zaleort.second_brain.Users.Domain.User.UserId
import com.zaleort.second_brain.Users.Domain.User.UserRepositoryInterface
import org.springframework.stereotype.Repository
import java.util.UUID

@Repository
class UserRepository(
    private val jpaUserRepository: JpaUserRepository
) : UserRepositoryInterface {
    override fun findById(id: UserId): User? {
        val uuid = UUID.fromString(id.value)
        val entity = jpaUserRepository.findById(uuid).orElse(null)
        return entity?.let {
            User.fromPrimitives(
                id = it.id.toString(),
                email = it.email,
                password = it.password
            )
        }
    }

    override fun findByEmail(email: UserEmail): User? {
        val entity = jpaUserRepository.findByEmail(email.value)
        if (entity == null) {
            return null
        }
        return entity.let {
            User.fromPrimitives(
                id = it.id.toString(),
                email = it.email,
                password = it.password
            )
        }
    }

    override fun save(user: User) {
        val userEntity = UserEntity(
            id = UUID.fromString(user.id.value),
            email = user.email.value,
            password = user.password.value
        )

        jpaUserRepository.save(userEntity)
    }

    override fun delete(id: UserId) {
        val uuid = UUID.fromString(id.value)
        val userEntity = jpaUserRepository.findById(uuid).orElse(null)
        if (userEntity == null) {
            throw Exception("User not found: " + id.value)
        }

        jpaUserRepository.delete(userEntity)
    }
}